widget/toolbar.go

```diff
@@ -13,31 +13,54 @@ type ToolbarItem interface {
 	ToolbarObject() fyne.CanvasObject
 }
 
+type toolbarActionButton struct {
+	Button
+
+	t *ToolbarAction
+}
+
+func newToolbarActionButton(t *ToolbarAction) *toolbarActionButton {
+	b := &toolbarActionButton{t: t}
+	b.ExtendBaseWidget(b)
+	return b
+}
+
+func (b *toolbarActionButton) Refresh() {
+	b.Icon = b.t.Icon
+	b.OnTapped = b.t.OnActivated
+
+	if b.t.Disabled {
+		b.Button.Disable()
+	} else {
+		b.Button.Enable()
+	}
+}
+
 // ToolbarAction is push button style of ToolbarItem
 type ToolbarAction struct {
 	Icon        fyne.Resource
 	OnActivated func() `json:"-"`
+	// Since: 2.5
+	Disabled bool
 }
 
 // ToolbarObject gets a button to render this ToolbarAction
 func (t *ToolbarAction) ToolbarObject() fyne.CanvasObject {
-	button := NewButtonWithIcon("", t.Icon, t.OnActivated)
-	button.Importance = LowImportance
-
-	return button
+	b := newToolbarActionButton(t)
+	b.Refresh()
+	return b
 }
 
 // SetIcon updates the icon on a ToolbarItem
 //
 // Since: 2.2
 func (t *ToolbarAction) SetIcon(icon fyne.Resource) {
 	t.Icon = icon
-	t.ToolbarObject().Refresh()
```

Comment:
You might be right - I think the expected behavior for SetIcon in the fyne API is to call Refresh internally, and not wait for the user to call refresh on the toolbar, right?

Reply:
Refresh should be called in a setter - so the state is drawn.
However calling it on the result of `ToolbarObject` won't do much as that is a new instance.

This line should be fixed not removed - feel free to address it or to leave it for now (but don't remove as that is also wrong).

Reply:
@andydotxyz You are right. At the moment, the only way to change anything in the `ToolbarAction` and have it visible in the `Toolbar` that the `ToolbarAction` has been assigned to is by calling the `Toolbar's` `Refresh` function. This will then re-render the `ToolbarAction` with the new state. This is a result of the change proposed here: https://github.com/fyne-io/fyne/pull/4347#discussion_r1375256904. Before that, I used to save a reference to the `Button` which is drawn on the screen in the `ToolbarAction`, which made it possible to update the state of the `Button` directly via the `ToolbarAction's` setters & properties.

With the current implementation I currently do not see any way to refresh the `ToolbarAction` if it has already been added to a `Toolbar`, other than calling the `Toolbar's` `Refresh` function. If we want to be able to update the `ToolbarAction` and immediately have the state reflected in the `Toolbars`, then I think it is necessary to save a reference to the rendered `Button` in the `ToolbarAction`, but this would go against the proposed change from https://github.com/fyne-io/fyne/pull/4347#discussion_r1375256904.

Let me know which approach you prefer so I can implement it.

Reply:
That is a good point, perhaps @dweymouth can comment so I don't contradict their thoughts.

Reply:
I think embedding the button is fine in this case as we have a reason to do so. My original thought was that we'd want a publicly settable `Disabled bool` field, but I see that I was mistaken as all the Disableable widgets do not export a public property- you only interact through the getters and setters.